### PR TITLE
Allison eng 659 opencode orchestrator troubleshooting tools

### PR DESCRIPTION
**Pull request by allisoneer**
## What problem(s) was I solving?

When orchestrating multiple OpenCode sessions, diagnosing stuck, silent, or failed sessions was difficult:
- No way to see session status (Idle/Busy/Retry) in the session list
- No way to identify which sessions the current orchestrator created vs pre-existing ones
- No tool to inspect detailed session state (pending messages, recent tool calls, retry info)
- Orchestrators had to guess why a session wasn't responding

## What user-facing changes did I ship?

**Enhanced `list_sessions` tool output** now includes:
- Session status (Idle, Busy, or Retry with attempt/reason details)
- `launched_by_you` marker to identify sessions this orchestrator created
- Working directory and file change statistics (additions/deletions/files)
- Both created and updated timestamps

**New `get_session_state` tool** for detailed session inspection:
- Current status with retry details (attempt count, reason, next retry time)
- Pending message count (user messages awaiting assistant reply)
- Last activity timestamp
- Recent tool calls with their states (pending/running/completed/error)

**New orchestrator documentation** with troubleshooting guidance explaining common diagnostic patterns.

## How I implemented it

1. **opencode_rs client enhancement**: Added `status_map()` method to batch-fetch all session statuses in one API call, avoiding N+1 queries when listing sessions.

2. **Session ownership tracking**: Added `spawned_sessions: Arc<RwLock<HashSet<String>>>` to the orchestrator server, populated when creating new sessions via the `run` tool.

3. **Enriched list_sessions**: The tool now fetches status map alongside session list, merges the data, and marks sessions as `launched_by_you` based on the tracked set.

4. **New get_session_state tool**: Fetches session details, status, and message history. Helper functions (`count_pending_messages`, `get_last_activity_time`, `extract_recent_tool_calls`) process the message history to extract diagnostic information.

5. **New types**: `SessionStatusSummary`, `ChangeStats`, `GetSessionStateInput/Output`, `ToolCallSummary`, `ToolStateSummary` with proper serde serialization and JSON schema derivation.

6. **Test coverage**: Added 9 new wiremock integration tests covering all status variants, launched_by_you tracking, message summarization, and error handling for unknown sessions.

## How to verify it

- [x] I ran the "check" and "test" recipes via the Just MCP tools (`tools_just_execute`) and they passed

## Description for the changelog

Add session troubleshooting diagnostics to the orchestrator: `list_sessions` now shows status (Idle/Busy/Retry) and ownership markers, new `get_session_state` tool provides detailed session inspection including pending messages and recent tool call states.

### CODE REVIEW

File: .opencode/command/openai.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+description: Interact with Linear for issue/project management
+agent: NormalOpenAI
+---
+$ARGUMENTS
```
[the remark this answers is not in the document]
Comment (allisoneer) [reply .opencode/command/openai.md:5]: 🤖 AI response: Addressed in pushed commit `85c8f22`. I updated `workflow.md` to document the `openai` command under the `NORMAL_OPENAI` command set and added the `NORMAL_OPENAI` agent variant alongside the existing workflow table entries.

File: .opencode/command/openai.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+description: Interact with Linear for issue/project management
```
[the remark this answers is not in the document]
Comment (allisoneer) [reply .opencode/command/openai.md]: 🤖 AI response: Addressed in pushed commit `85c8f22`. `.opencode/command/openai.md` now uses the correct OpenAI-focused description metadata instead of the stale Linear text.

File: apps/opencode-orchestrator-mcp/src/tools.rs [hunk 3]
```diff
@@ -1016,6 +1068,184 @@ impl Tool for ListSessionsTool {
...
+                        message,
+                        next,
+                    },
+                    Ok(SessionStatusInfo::Idle) | Err(_) => SessionStatusSummary::Idle,
+                };
+
+                let messages = client.messages().list(session_id).await.unwrap_or_default();
+                let pending_message_count = count_pending_messages(&messages);
+                let last_activity = get_last_activity_time(&messages);
+                let recent_tool_calls = extract_recent_tool_calls(&messages, 10);
```
[the remark this answers is not in the document]
Comment (allisoneer) [reply apps/opencode-orchestrator-mcp/src/tools.rs]: 🤖 AI response: Addressed in pushed commit `85c8f22`. `get_session_state` now propagates `/session/status` and message-list fetch failures as `ToolError::Internal` instead of fabricating `Idle` status or empty diagnostics, and the failure-path tests were updated accordingly.

File: .opencode/orchestrator_sysprompt_gpt54.md [hunk 4]
```diff
@@ -209,6 +209,37 @@ For multi-session implementations after summarization:
...
+## Session Troubleshooting
+
+When sessions appear stuck, fail silently, or return unexpected results, use these diagnostic tools:
+
+1. `orchestrator_list_sessions` — Lists all sessions with:
+   - Session status (Idle/Busy/Retry)
+   - `launched_by_you` marker for sessions you created
+   - Working directory and change statistics
+
+2. `orchestrator_get_session_state` — Detailed inspection of a specific session (surfaced prompt alias for the MCP app's `get_session_state` tool):
```
[the remark this answers is not in the document]
Comment (allisoneer) [reply .opencode/orchestrator_sysprompt_gpt54.md:239]: 🤖 AI response: Addressed in pushed commit `85c8f22`. I added `orchestrator_get_session_state` to the `<tool_definitions>` section in `.opencode/orchestrator_sysprompt_gpt54.md`, including its `session_id` parameter.

File: apps/opencode-orchestrator-mcp/src/tools.rs [hunk 5]
```diff
@@ -977,15 +992,52 @@ impl Tool for ListSessionsTool {
                     server.client().sessions().list().await.map_err(|e| {
                         ToolError::Internal(format!("Failed to list sessions: {e}"))
                     })?;
+                let status_map = server
+                    .client()
+                    .sessions()
+                    .status_map()
+                    .await
+                    .unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (allisoneer) [reply apps/opencode-orchestrator-mcp/src/tools.rs]: 🤖 AI response: Addressed in pushed commit `85c8f22`. `list_sessions` no longer falls back to `Idle` for every session when `status_map()` fails; it now keeps `status` as `None`, which matches the existing output type and `unknown` rendering path.

File: .opencode/orchestrator_sysprompt.md [hunk 6]
```diff
@@ -213,6 +213,22 @@ Limit responses to 4 bullets maximum, 2 sentences each. When reporting session r
 
 **Directory access requests:** These often indicate a session is doing something incorrectly. Sessions should use `ask_agent` with `location=references` to explore reference repos, not direct file access. Reject directory permission requests and redirect the session to use the appropriate agent tools.
 
+### Session Troubleshooting
+
+When a session appears stuck, fails silently, or returns unexpected results:
+
+1. **List all sessions** using `orchestrator_list_sessions` to see session status (Idle/Busy/Retry) and identify which sessions you launched.
+2. **Inspect detailed state** using `orchestrator_get_session_state` (surfaced prompt alias for the MCP app's `get_session_state` tool) with the session ID to see:
```
[the remark this answers is not in the document]
Comment (allisoneer) [reply .opencode/orchestrator_sysprompt.md:223]: 🤖 AI response: Addressed in pushed commit `85c8f22`. I added `orchestrator_get_session_state` to the orchestrator tool inventory in `.opencode/orchestrator_sysprompt.md` with a thread-specific description of the status, pending-message, recent-tool-call, and last-activity inspection it provides.

File: apps/opencode-orchestrator-mcp/src/tools.rs [hunk 7]
```diff
@@ -1016,6 +1068,184 @@ impl Tool for ListSessionsTool {
...
+            if !last_was_user {
+                pending += 1;
+            }
+            last_was_user = true;
+        } else if message.role() == "assistant" {
+            break;
+        }
+    }
+
+    pending
```
[the remark this answers is not in the document]
Comment (allisoneer) [reply apps/opencode-orchestrator-mcp/src/tools.rs:1079]: 🤖 AI response: Addressed in pushed commit `85c8f22`. `count_pending_messages()` now increments for each trailing user message, and the wiremock coverage was extended with a case that asserts two trailing user messages produce a count of `2`.

File: apps/opencode-orchestrator-mcp/src/tools.rs [hunk 8]
```diff
@@ -1016,6 +1068,184 @@ impl Tool for ListSessionsTool {
...
+                    call_id: call_id.clone(),
+                    tool_name: tool.clone(),
+                    state,
+                    started_at,
+                    completed_at,
+                });
+
+                if tool_calls.len() >= limit {
+                    return tool_calls;
+                }
```
[the remark this answers is not in the document]
Comment (allisoneer) [reply apps/opencode-orchestrator-mcp/src/tools.rs:1135]: 🤖 AI response: Investigated, but not changing this in the `85c8f22` batch. The current implementation has demonstrated recent-first ordering across messages, but our existing coverage only exercises one tool part per message; we do not have a concrete failing case for within-message multi-tool-part ordering yet, so I’m not changing `message.parts` iteration in this batch without that evidence.

File: crates/services/opencode-rs/src/http/sessions.rs [hunk 9]
```diff
@@ -133,6 +134,31 @@ impl SessionsApi {
...
+            SessionStatusResponse::Map(map) => map,
+            SessionStatusResponse::Legacy(summary) => {
+                let mut map = HashMap::new();
+                if summary.busy
+                    && let Some(active_session_id) = summary.active_session_id
+                {
+                    map.insert(active_session_id, SessionStatusInfo::Busy);
+                }
+                map
+            }
```
[the remark this answers is not in the document]
Comment (allisoneer) [reply crates/services/opencode-rs/src/http/sessions.rs]: 🤖 AI response: Investigated, but not changing this in the `85c8f22` batch. Legacy `/session/status` responses only deserialize into `busy` plus `active_session_id`, so retry metadata is not available to preserve on that path. Modern per-session map responses already preserve `Retry` details; the legacy normalization branch can only represent `Busy`/`Idle` with the current response shape.

File: apps/opencode-orchestrator-mcp/src/tools.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (allisoneer) [reply apps/opencode-orchestrator-mcp/src/tools.rs:1135]: Hey, human here, How can we get a failing test to make sure we're aligned and buttoned up on how this is is working? If we ask a model to call tools in parallel (as many of our workflows do do), then wouldn't that be mutliple tool calls in a single assistant response? This seems like something we would/could be able to actually test? It sounds potentially quite wise to look into this? right? I'm giving you (not you coderabbit, the other agent who will read this) permission to expand scope within this branch to at the very least properly investigate this, if not make changes/add tests to ensure we're clean and idiomatic here.

File: crates/services/opencode-rs/src/http/sessions.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (allisoneer) [reply crates/services/opencode-rs/src/http/sessions.rs]: Hey friends (coderabbit and the agent who will read this), we actually don't care about supporting old versions of old code. If version 1.3.17 doesn't need to support something, then we should remove that code! The key thing here will be to verify/confirm what the 1.3.17 branch actually supports and to be diligent/deliberate about making sure our codebase is simply supporting what that source code is doing. No action needed from you code rabbit, this message is for our agent friend.

File: .opencode/orchestrator_sysprompt.md [hunk 10]
```diff
@@ -213,6 +214,22 @@ Limit responses to 4 bullets maximum, 2 sentences each. When reporting session r
...
+   - Current status including retry information
+   - Pending message count
+   - Recent tool calls and their states (pending/running/completed/error)
+   - Last activity timestamp
+3. **Common patterns**:
+   - Session stuck in "Busy" for too long → may indicate a hung tool or deadlock
+   - Session in "Retry" → provider overload or rate limiting; check retry details
+   - Tool calls stuck in "pending" or "running" → execution interrupted or timed out
+   - `launched_by_you: false` → session was created by another process; may need context
```
[the remark this answers is not in the document]
Comment (allisoneer) [reply .opencode/orchestrator_sysprompt.md]: We should check to make sure our prompts match the actual initial desired intentions expressed in our ticket.md in the thoughts artifact as well as the current codebase. Thanks for calling this out coderabbit, this message is just steering/instructions for our agent.

File: workflow.md [hunk 11]
```diff
@@ -33,22 +33,28 @@ Inside of normal we have a handful of tools, most are custom, and pluggable into
...
 
 ```
 NORMAL:     research, create_plan_init, create_plan_final, implement_plan,
             review_pr_comments, init
+NORMAL_OPENAI:
+            openai, research_openai, create_plan_init_openai,
+            create_plan_final_openai, implement_plan_openai,
+            review_pr_comments_openai, capture_pr_comments_openai,
+            resolve_pr_comments_openai, resume_work_openai,
+            unwind_openai, decide_findings_openai, frame_openai
```
[the remark this answers is not in the document]
Comment (allisoneer) [reply workflow.md]: Yes! Great callout! WE should probably update it with all sorts of new agents and information and such. workflow.md likely deserves being updated fully with an entire research investigation into anything/everything that might be missing thus far.

[the remark this answers is not in the document]
Comment (allisoneer) [reply workflow.md]: 🤖 AI response: [AI] Fixed in `fd5c6f2`.

I synced `workflow.md` Level 0 with the actual orchestrator tool surface by adding `orchestrator_get_session_state` and `orchestrator_respond_question`. I also kept the prompt/tooling docs aligned in the same batch. Verified with `just check` and `just test`.

File: .opencode/orchestrator_sysprompt.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (allisoneer) [reply .opencode/orchestrator_sysprompt.md]: 🤖 AI response: [AI] Fixed in `fd5c6f2`.

I updated both orchestrator prompts to account for `unknown` session status from `orchestrator_list_sessions` and added troubleshooting guidance to retry/investigate that case instead of treating it like a normal `Idle` state. Verified with `just check` and `just test`.

File: crates/services/opencode-rs/src/http/sessions.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (allisoneer) [reply crates/services/opencode-rs/src/http/sessions.rs]: 🤖 AI response: [AI] Fixed in `fd5c6f2`.

I checked pinned upstream OpenCode `v1.3.17` and it serves `/session/status` as the per-session map shape, not the old `{busy, activeSessionId}` summary. Based on that target-version contract and the maintainer follow-up here, I removed the unnecessary local legacy compatibility branch and updated the affected tests. Verified with `just check` and `just test`.

File: apps/opencode-orchestrator-mcp/src/tools.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (allisoneer) [reply apps/opencode-orchestrator-mcp/src/tools.rs:1135]: 🤖 AI response: [AI] Fixed in `fd5c6f2`.

I added a focused integration test for one assistant message containing two tool parts, then reversed within-message part iteration in `extract_recent_tool_calls()` so `recent_tool_calls` now satisfies the documented "most recent first" contract for that case. Verified with `just check` and `just test`.

File: crates/services/opencode-rs/src/http/sessions.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (allisoneer) [reply crates/services/opencode-rs/src/http/sessions.rs]: 🤖 AI response: Checked the current branch and pinned target directly. This branch is pinned to OpenCode `1.3.17` (`crates/services/opencode-rs/src/version.rs:9-10`), upstream status for that target is map-based, and the current code already parses `/session/status` as `HashMap<String, SessionStatusInfo>` in `status()`, `status_for()`, and `status_map()` (`crates/services/opencode-rs/src/http/sessions.rs:115-173`). The old legacy input-normalization path is already gone, so no further behavior change is needed for this thread.

File: apps/opencode-orchestrator-mcp/src/tools.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (allisoneer) [reply apps/opencode-orchestrator-mcp/src/tools.rs:1135]: 🤖 AI response: Checked the current branch directly. `extract_recent_tool_calls()` already iterates both `messages` and `message.parts` in reverse, so recent tool calls are already returned recent-first across messages and within a message (`apps/opencode-orchestrator-mcp/src/tools.rs:1089-1135`). There is also already a within-message ordering test asserting `call-newer` before `call-older` (`apps/opencode-orchestrator-mcp/tests/list_tools_wiremock.rs:443-518`), so no production change is needed for this thread. Optional follow-up if we want stronger regression coverage: add a 3+ tool-part and/or limit-boundary test.

File: workflow.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (allisoneer) [reply workflow.md]: 🤖 AI response: Checked the current docs directly. `workflow.md` Level 0 now includes `orchestrator_get_session_state` in the orchestrator tool inventory (`workflow.md:13-22`).

File: .opencode/orchestrator_sysprompt.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (allisoneer) [reply .opencode/orchestrator_sysprompt.md]: 🤖 AI response: Checked the current prompt directly. The troubleshooting section now explicitly handles sessions shown as `unknown` and says to retry or investigate instead of treating that state like `Idle` (`.opencode/orchestrator_sysprompt.md:218-233`).

File: crates/agentic-tools/core/src/schema.rs [hunk 12]
```diff
@@ -404,15 +353,15 @@ mod tests {
...
-            // AddNullable should add "nullable": true
+            // Option<String> should produce {"type": ["string", "null"]}
             assert_eq!(
-                a.get("nullable"),
-                Some(&serde_json::Value::Bool(true)),
-                "Option<T> fields should have nullable: true"
+                a.get("type"),
+                Some(&serde_json::json!(["string", "null"])),
+                "Option<T> fields should have type array with null"
             );
```
[the remark this answers is not in the document]
Comment (allisoneer) [reply crates/agentic-tools/core/src/schema.rs]: 🤖 AI response: Fixed in `eb13e2d`. The `Option<String>` assertions are now order-insensitive in `crates/agentic-tools/core/src/schema.rs:357-368` and `crates/agentic-tools/core/src/schema.rs:502-519`, so the tests no longer depend on a stable `type` array ordering.

File: crates/agentic-tools/core/src/schema.rs [hunk 13]
```diff
@@ -172,19 +172,17 @@ impl SchemaEngine {
...
 
-/// Centralized schema generation using Draft 2020-12 with AddNullable transform.
+/// Centralized schema generation using Draft 2020-12.
 ///
-/// This module provides cached schema generation matching the MCP Rust SDK pattern:
+/// This module provides cached schema generation for MCP:
 /// - JSON Schema Draft 2020-12 (MCP protocol requirement)
-/// - AddNullable transform for `Option<T>` fields
+/// - `Option<T>` fields generate `{"type": ["T", "null"]}` (schemars default)
 /// - Thread-local caching keyed by TypeId for performance
```
[the remark this answers is not in the document]
Comment (allisoneer) [reply crates/agentic-tools/core/src/schema.rs:181]: 🤖 AI response: Fixed in `eb13e2d`. The doc comment is now narrowed to say `Option<T>` includes `null` in schema shape, with `type` arrays for simple scalar cases and `anyOf`/`$ref` forms for complex types, in `crates/agentic-tools/core/src/schema.rs:175-181`.

File: apps/opencode-orchestrator-mcp/src/tools.rs [hunk 14]
```diff
@@ -977,15 +992,46 @@ impl Tool for ListSessionsTool {
...
+                                    attempt,
+                                    message,
+                                    next,
+                                } => SessionStatusSummary::Retry {
+                                    attempt: *attempt,
+                                    message: message.clone(),
+                                    next: *next,
+                                },
+                            })
+                        });
```
[the remark this answers is not in the document]
Comment (allisoneer) [reply apps/opencode-orchestrator-mcp/src/tools.rs]: 🤖 AI response: Fixed in `99476e7`. `list_sessions` now treats a successful `status_map()` lookup with no entry as `Idle` in `apps/opencode-orchestrator-mcp/src/tools.rs:1003-1020`, and the empty-status-map wiremock case now asserts `Idle` for both sessions in `apps/opencode-orchestrator-mcp/tests/list_tools_wiremock.rs:179-195`.

File: workflow.md [hunk 15]
```diff
@@ -33,22 +35,28 @@ Inside of normal we have a handful of tools, most are custom, and pluggable into
...
+|---------------|--------------------|---------------|---------------|-----------------------|-----------------------|
+| File ops      | File ops           | File ops      | File ops      | File ops              | Read only             |
+| Search        | Search             | Search        | Search        | Search                | Search (cli_*)        |
+| Tasks         | Tasks              | Tasks         | Tasks         | Tasks                 |                       |
+| Just runner   | Just runner        | Just runner   | Just runner   | Just runner           | Just runner (limited) |
+| Thoughts      | Thoughts           | Thoughts      | Thoughts      | Thoughts              | Thoughts (write only) |
+| GitHub PRs    | GitHub PRs         | GitHub PRs    | GitHub PRs    | GitHub PRs            |                       |
+| Sub-agents    | Sub-agents         | Sub-agents    | Sub-agents    | Sub-agents            | Reasoning model       |
+|               | GPT-5.4-oriented commands | + mcp_bash (shell) | + 9 Linear tools | + 18 Browser automation | + review_* tools |
```
[the remark this answers is not in the document]
Comment (allisoneer) [reply workflow.md]: 🤖 AI response: Fixed in `9193c81`. `workflow.md` now shows Playwright as `41` total tools with `+ 22 Browser automation` in `workflow.md:38-47`, matching the active orchestrator prompt's Playwright expansion guidance in `.opencode/orchestrator_sysprompt.md:239-247`.

File: apps/opencode-orchestrator-mcp/src/tools.rs [hunk 16]
```diff
@@ -1016,6 +1065,184 @@ impl Tool for ListSessionsTool {
...
+                    },
+                    SessionStatusInfo::Idle => SessionStatusSummary::Idle,
+                };
+
+                let messages = client.messages().list(session_id).await.map_err(|e| {
+                    ToolError::Internal(format!("Failed to list messages: {e}"))
+                })?;
+                let pending_message_count = count_pending_messages(&messages);
+                let last_activity = get_last_activity_time(&messages);
+                let recent_tool_calls = extract_recent_tool_calls(&messages, 10);
```
[the remark this answers is not in the document]
Comment (allisoneer) [reply apps/opencode-orchestrator-mcp/src/tools.rs:1212]: 🤖 AI response: Fixed in `5e10d8e`.

`get_session_state` now passes `session.time.updated` into `get_last_activity_time`, and `get_last_activity_time` now uses that fallback when the message list is empty, so history-empty sessions still report last activity from session metadata (`apps/opencode-orchestrator-mcp/src/tools.rs:1082-1092`, `apps/opencode-orchestrator-mcp/src/tools.rs:1208-1211`). I also added a regression test for the empty-message case (`apps/opencode-orchestrator-mcp/src/tools.rs:1730-1733`). Verified with `just check` and `just test`.
